Adds --prefix flag to the kubectl log command

m1kola opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

One of the sig-cli meetings during discussion of #67573 it was suggested to add a new flag (`--prefix`) which will prepend log lines with information about log sources. It makes a lot of sense when reading or following logs from multiple sources.

Read logs from multiple sources:

```
kubectl logs -lapp=some-app --all-containers --prefix
```

Follow logs from multiple sources:

```
kubectl logs -lapp=some-app --all-containers --prefix --follow
```

**Does this PR introduce a user-facing change?**:

```release-note
A new `--prefix` flag added into `kubectl logs` which prepends each log line with information about it's source (pod name and container name)
```

---

/sig cli

neolit123 commented on pkg/kubectl/cmd/logs/logs.go:
```diff
@@ -64,6 +64,10 @@ var (
 		kubectl logs -f -lapp=nginx --all-containers=true
 
+		# Begin streaming the logs from all containers in pods defined by label app=nginx.
+		# Each line will be prefixed with logs source (pod name and container name)
```
`with logs source` -> `with the log source`

neolit123 commented on pkg/kubectl/cmd/logs/logs.go:
```diff
@@ -161,6 +166,7 @@ func NewCmdLogs(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.C
 	cmdutil.AddPodRunningTimeoutFlag(cmd, defaultPodLogsTimeout)
 	cmd.Flags().StringVarP(&o.Selector, "selector", "l", o.Selector, "Selector (label query) to filter on.")
 	cmd.Flags().IntVar(&o.MaxFollowConcurency, "max-log-requests", o.MaxFollowConcurency, "Specify maximum number of concurrent logs to follow when using by a selector. Defaults to 5.")
+	cmd.Flags().BoolVar(&o.Prefix, "prefix", o.Prefix, "Prefix each log line with log source (pod name and container name)")
```
`with the log source`

neolit123 commented on pkg/kubectl/polymorphichelpers/interface.go:
```diff
@@ -28,8 +28,16 @@ import (
 	"k8s.io/kubernetes/pkg/kubectl"
 )
 
+// LogsForObjectResponseWrapper defines an interface for the response struct that allows to get access to source pod and container for logs
```
`access to source` -> `access to the source`

neolit123 reviewed (commented):
thanks, added minor comments.

draft PRs are not well supported by test-infra yet.
(promoting the PR to non-draft and adding WIP in the title is the alternative)

there is a typo in the commit and PR title:
`--prefex` -> `--prefix`.

neolit123 commented:
/priority backlog
there seems to be a panic in the bazel-test job:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/76471/pull-kubernetes-bazel-test/1116473315288944641/

m1kola commented:
@neolit123 thanks for the feedback. I'll update as soon as I can.

I'll also be looking into fixing panic in tests. It's still work in progress ;)

m1kola commented on pkg/kubectl/polymorphichelpers/logsforobject.go:
```diff
@@ -68,10 +68,37 @@ func logsForObjectWithClient(clientset corev1client.CoreV1Interface, object, opt
 		if !allContainers {
-			return []rest.ResponseWrapper{clientset.Pods(t.Namespace).GetLogs(t.Name, opts)}, nil
+			var container *v1.Container
+			if opts == nil || len(opts.Container) == 0 {
```
This else-if keeps the `kubectl` behaviour the same to a user as in previous versions, but the output will be a bit different. For example, for `kubectl -n logging-test logs -lapp=logging_test` where `-lapp=logging_test` matches a pod with more than 1 container will be:

```
Error from server (BadRequest): a container name must be specified for pod logging-test-be-7657b857c7-4b9jr, choose one of: [logging-container-1 logging-container-2]
```

New output:

```
error: a container name must be specified for pod logging-test-be-7657b857c7-4b9jr, choose one of: [logging-container-1 logging-container-2]
```

Same when provided container name is incorrect (`kubectl -n logging-test logs -lapp=logging_test -c test`):

```
error: container test is not valid for pod logging-test-be-7657b857c7-4b9jr
```

instead of 

```
Error from server (BadRequest): container test is not valid for pod logging-test-be-7657b857c7-4b9jr
```

It happens because now we check pod object on the client before sending request to get logs.

m1kola replied:
It works, but I'm not very happy about this approach for several reasons:

* On client I duplicate some of the logic we already have on server (see [`LogLocation ` from `pkg/registry/core/pod/strategy.go`](https://github.com/kubernetes/kubernetes/blob/18b4e1b84c241b89392a37721607d0b19382b92f/pkg/registry/core/pod/strategy.go#L286-L361))
* `LogsForObjectResponseWrapper` looks a bit clumsy

Any suggestions are very welcome.

neolit123 replied:
> On client I duplicate some of the logic we already have on server (see LogLocation from pkg/registry/core/pod/strategy.go)

ideally there should be a common backend.
although i'm not sure where it should be placed.

> LogsForObjectResponseWrapper looks a bit clumsy

possibly change the function name to better outline what it does.
e.g. SourceObjectResponseWrapper
can't think of good names ATM.

soltysh replied:
> It works, but I'm not very happy about this approach for several reasons:

Ditto, I'd imagine this more like a writer wrapper in `kubectl logs` and on the backend I'd be inclined to propose to modify `LogsForObjectFunc` so that it returns `map[corev1.ObjectReference]rest.ResponseWrapper` instead of current `[]rest.ResponseWrapper` wdyt? The reason why I'm going with `corev1.ObjectReference` is that this does not restrict us into the format of the eventual string we'd put there and will allow consumers to format it per their own taste.

m1kola commented:
/retest

m1kola commented:
/retest

m1kola commented:
@neolit123 could you please take another look? I finalised the PR and addressed your feedback.

Please, ignore the latest tests failures - tests failed for a weird reason so I had to run retest and I'm pretty sure they will pass soon because the only thing I changed since the last successful test run was the commit message (fixing the typo: `prefex` -> `prefix`).

neolit123 reviewed (commented):
@m1kola 
thanks for the update.

at this point i would ping the kubectl maintainers for more eyes on the change.
you can link this PR in the #sig-cli slack channel or bringing it in the SIG CLI meeting agenda.
https://github.com/kubernetes/community/tree/master/sig-cli
:+1:

soltysh commented on pkg/kubectl/cmd/logs/logs.go:
```diff
@@ -64,6 +64,10 @@ var (
 
+		# Begin streaming the logs from all containers in pods defined by label app=nginx.
+		# Each line will be prefixed with the log source (pod name and container name)
+		kubectl logs -f -lapp=nginx --all-containers=true --prefix
```
I'd drop this example, we don't need to have examples for every possible combination. You could add a single line of comment to the previous about `--prefix` possibility. I'm worried that at some point in time this help page will be longer than single page which makes it hard to read, see `kubectl run --help` as an example of what I don't want to have.

m1kola commented:
@soltysh I updated the PR. Please take another look.

Commit:
Adds --prefix flag to the kubectl log command

m1kola commented:
Restarting the job. It seems to be fixed in https://github.com/kubernetes/test-infra/pull/14203

/test pull-kubernetes-node-e2e

m1kola commented:
@soltysh could you please take another look?

I rebased it on top of the `master` branch and added support for ephemeral containers (https://github.com/kubernetes/enhancements/issues/277).

/remove-lifecycle rotten

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:
```diff
@@ -344,16 +348,39 @@ func (o LogsOptions) parallelConsumeRequest(requests []rest.ResponseWrapper) err
+	// including a container name (not an index) so we can get a container name
+	// without making an extra API request.
+	var containerName string
+	re := regexp.MustCompile(`spec\.(?:initContainers|containers|ephemeralContainers){(.+)}`)
```
regexp compile is expensive, let's compile this only once and re-use.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:
```diff
@@ -344,16 +348,39 @@ func (o LogsOptions) parallelConsumeRequest(requests []rest.ResponseWrapper) err
+		containerName = containerNameMatches[1]
+	}
+
+	prefix := fmt.Sprintf("[pod/%s -c %s] ", ref.Name, containerName)
```
`pod/<pod-name>/<container-name>` would be the best imo.

m1kola replied:
@soltysh  I can do that. The reason why I decided to go with `pod/%s -c %s` is becasue it's easy to copy-paste this bit into a new `kubectl logs` call (in a separate window). It was very handy while working on the PR itself.

But I do not have strong preferences here.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:
```diff
@@ -384,3 +411,25 @@ func DefaultConsumeRequest(request rest.ResponseWrapper, out io.Writer) error {
+
+	// Perform an "atomic" write of a prefix and p to make sure that it doesn't interleave
+	// sub-line when used concurrently with io.PipeWrite.
+	n, err := pw.writer.Write(append(pw.prefix, p...))
```
I'd just 
```
return pw.writer.Write(append(pw.prefix, p...))
```

m1kola replied:
From this function we must return number of bytes (`n`) written from `p`, where `n` must be `0 <= n <= len(p)`to meet the `io.Writer` interface requirements.

`return pw.writer.Write(append(pw.prefix, p...))` will be violating the `io.Writer` interface: it can return `n > len(p)` because we write `p` and `pw.prefix`. Other packages/functions will misbehave if we do that (I had issues with `io.Copy` which compares number of bytes internally, for example).

soltysh replied:
Roger that :+1:

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:
```diff
@@ -344,16 +348,39 @@ func (o LogsOptions) parallelConsumeRequest(requests []rest.ResponseWrapper) err
 }
 
+func (o LogsOptions) addPrefixIfNeeded(ref corev1.ObjectReference, writer io.Writer) io.Writer {
+	if !o.Prefix {
```
I'd check if the ref is not empty one too.

soltysh reviewed (changes requested):
A few more comments.

soltysh reviewed (approved):
/lgtm
/approve

apelisse commented:
I can't tell if it's because of that PR, but there is a test that flakes about 5% of the time:
```
$ bazel test --runs_per_test=1000 //staging/src/k8s.io/kubectl/pkg/cmd/logs:go_default_test
...
//staging/src/k8s.io/kubectl/pkg/cmd/logs:go_default_test                FAILED in 50 out of 1000 in 0.3s
...
```
